src/Adapter/Shop/Url/ProductProvider.php

```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Shop\Url;
+
+use Link;
+use PrestaShop\PrestaShop\Core\Shop\Url\UrlProviderInterface;
+
+/**
+ * Class ProductProvider provides base Front Office URL for context shop.
+ */
+class ProductProvider implements UrlProviderInterface
+{
+    /**
+     * @var Link
+     */
+    private $link;
+
+    /**
+     * @param Link $link
+     */
+    public function __construct(Link $link)
+    {
+        $this->link = $link;
+    }
+
+    /**
+     * Create a link to a product.
+     *
+     * @param int $productId
+     * @param string $rewrite
+     *
+     * @return string
+     */
+    public function getUrl($productId = null, $rewrite = null)
```

Comment:
```suggestion
    public function getUrl(int $productId = null, string $rewrite = null): string
```

Reply:
I'm not sure how PHPStan will react to this, because this method already doesn't respect the implemented Interface anyway 😅